### PR TITLE
Article: ExpressJS

### PR DESCRIPTION
**Pull request by rohanreddych**
Description
> Express is a lightweight framework on top of Node, it adds functionality like middleware, template engines and routing

### CODE REVIEW

**Comment by hectorkambow** [comment]
plagiarism check done

File: articles/express/index.md [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+
+When creating a Node.js web application, we write a single JavaScript application which listens to requests from the browser, based on the request the function will send back some data or an HTML web page.
+
+![flow of a request](/engineering-education/express/node.png)
+
+A *request handler* is a JavaScript function which takes a request and sends an appropriate response.
+
+Node.js APIs can get complex and writing how to handle a single request can end up being to 50 lines of code. Express makes it easier to write web applications with Node.js.
+
+![flow with express](/engineering-education/express/express.png)
```
Comment (hectorkambow) [inline articles/express/index.md:30]: Just wanted to double check if the images in the article needed sources.. 🚀

**Comment by hectorkambow** [comment]
@louisefindlay23 - I made some changes during the preliminary review - the article is good to be tech reviewed.
Feel free to suggest any changes you see fit - 🚀

File: articles/express/index.md [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+
+When creating a Node.js web application, we write a single JavaScript application which listens to requests from the browser, based on the request the function will send back some data or an HTML web page.
+
+![flow of a request](/engineering-education/express/node.png)
+
+[Image Credit](https://www.manning.com/books/express-in-action)
+
+A *request handler* is a JavaScript function which takes a request and sends an appropriate response.
+
+Node.js APIs can get complex and writing how to handle a single request can end up being to 50 lines of code. Express makes it easier to write web applications with Node.js.
```
Comment (louisefindlay23) [inline articles/express/index.md:30]: being to doesn't make sense. Should it be over? Consider rephrasing to easier to write **Node.js web applications** to improve flow.

File: articles/express/index.md [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+
+![flow with express](/engineering-education/express/express.png)
+
+[Image Credit](https://www.manning.com/books/express-in-action)
+
+### Advantages of using Express with Node.js
+- Express lets you take away a lot of Node.js complexities while adding helpful functions to a Node.js HTTP server.
+- Instead of a large request handler function, Express allows us to handle requests by writing many small modular and maintainable functions.
+- Express is *not opinionated*, meaning Express does not enforce any "right way" of doing things. You can use any compatible middleware, you can structure the app as you wish making it flexible.
```
Comment (louisefindlay23) [inline articles/express/index.md:39]: You should replace the comma after middleware with and.

File: articles/express/index.md [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+- We can integrate with a [view rendering engine](https://www.digitalocean.com/community/tutorials/nodejs-express-template-engines) of our choice like Jade, Pug, EJS, etc. A template engine enables you to use static template files and at runtime change the values of variables in those files.
```
Comment (louisefindlay23) [inline articles/express/index.md:40]: You should explain that a view rendering engine and template engine are one and the same. Beginners may be confused.

File: articles/express/index.md [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+  res.send("HELLO WORLD");
+  });
+
+//3
+app.listen(3000, function(){
+  console.log("Application started and Listening on port 3000");
+  });
+```
+
+The code above creates a basic express application. To run this script, go to your command prompt and enter the command `node app.js` in the project directory. In the console, we can see `Application started and Listening on port 3000` and if we visit `http://localhost:3000/` we can see `HELLO WORLD`.
```
Comment (louisefindlay23) [inline articles/express/index.md:68]: Express should have a capital A

File: articles/express/index.md [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+
+Let's look at what the code above does.
+
+The first line imports the express module. The second line creates an Express application by calling the top-level `express()` function. Our `app` variable (express application) has methods for handling requests and configuring how the application behaves. We can create multiple apps this way, each with their own requests and responses.
+
+The code under 2 `app.get()` is a function, called *route definition*, which tells the express app how to handle an HTTP `GET` request to our server. This function takes 2 main parameters, the first is the route or path which is the relative path from the root of the server; the second is a function that is invoked whenever there is a request to that path. In this case, we are listening for `GET` requests to `/` which is the root of the website.
```
Comment (louisefindlay23) [inline articles/express/index.md:74]: The code under 2 would be better references as the second part of code.

File: articles/express/index.md [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+Express comes with **`express.static`** middleware bundled with it, it can be used to serve static files instead of the function in the previous section. It provides better security and performance than the function we wrote.
+
+`JavaScript
+app.use(express.static("static") //relative path
+`
+
+Any requested files in the directory "static" are served. `localhost:3000/dummy_file.txt` will show the same result as above. We can call `static()` multiple times. If the file is not in the one directory, then the request is passed onto the next `static()` call. When your app gets a request like `/hello.html`, it returns the hello.html file in the public directory.
+
+### Routing
+Express makes request handling easier by mapping requests to different request handlers. A request handler is a function which handles all the requests to a specific path with a specific HTTP verb. In the basic example above, we saw how to handle a `GET` request. As an application grows in size the routes grow as well as do the request handlers.
```
Comment (louisefindlay23) [inline articles/express/index.md:185]: HTTP method.

File: articles/express/index.md [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+</html>
+```
+If you go to the webpage you can see `Hello and Welcome !!!`. You can write JavaScript expressions inside the `<%= exp %>`. Look at the [docs](https://ejs.co/#docs) for the complete syntax and rules.
+
+**Testing Express Applications**.
+
+Testing is an important part of developing software. Read [this post](/engineering-education/node-testing/) where we discuss testing Node.js applications using Mocha and Chai.
+
+### Conclusion
+The minimalistic philosophy of Express may not be suited for everyone's needs, because you can make mistakes while you are making those decisions about your applications infrastructure. One of the best practice is to have a directory structure like this or something similar to this for your express app
```
Comment (louisefindlay23) [inline articles/express/index.md:291]: Capitialise E in Express.

File: articles/express/index.md [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+└── routes/
+|   ├── api.js
+|   └── router.js
+├── app.js
+├── package.json
+├── tests/
+└── bin/
+```
+
+Express is an unopinionated framework which works best by making use of third party software to create a full-fledged application. Simplicity is a high level goal for software in general, when writing Express applications keep this in mind.
```
Comment (louisefindlay23) [inline articles/express/index.md:310]: Consider linking to other Node.js Section articles that are related to your article. Some mention testing, EJS, Node.js best practises and developing a Node.js app using Express.

**Review by louisefindlay23 — CHANGES_REQUESTED** [review]
Hi, @rohanreddych. This is a great article. I certainly learned things about Express, I didn't know before such as router and using regular expression in routes. 

I picked up some small corrections you should make (both technical and grammatical) and suggested some places where you could improve readability/flow before I test all your code.

**Comment by rohanreddych** [comment]
@louisefindlay23 I addressed your concerns, let me know if any changes are needed.

File: articles/express/index.md [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+author: rohan-reddy
+date: 2020-09-18T00:00:00-11:00
+topics: [Node.js]
+excerpt_separator: <!--more-->
+images:
+
+  - url: /engineering-education/express/hero.jpg
+    alt: expressjs example text image
+---
+[Node.js](https://www.section.io/engineering-education/history-of-nodejs/) is a JavaScript run time environment which is used to create server-side applications and tools. Node.js is fast, portable and written in JavaScript. Node.js does not directly support common tasks such as handling requests, serving files, handling HTTP methods such as `GET` and `POST` this is where Node.js's [rich ecosystem](https://www.section.io/engineering-education/most-useful-nodejs-packages/) comes to our aid.
```
Comment (louisefindlay23) [inline articles/express/index.md:17]: Add but before this

File: articles/express/index.md [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+<!--more-->
+
+**Express.js (Express) is a light web framework which sits on top of Node.js and it adds functionality ([middleware](https://expressjs.com/en/guide/writing-middleware.html) , [routing](https://en.wikipedia.org/wiki/Routing), etc.) and simplicity to Node.js.**
```
Comment (louisefindlay23) [inline articles/express/index.md:20]: Remove space between .html) and ,

File: articles/express/index.md [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+
+![flow with express](/engineering-education/express/express.png)
+
+[Image Credit](https://www.manning.com/books/express-in-action)
+
+### Advantages of using Express with Node.js
+- Express lets you take away a lot of Node.js complexities while adding helpful functions to a Node.js HTTP server.
+- Instead of a large request handler function, Express allows us to handle requests by writing many small modular and maintainable functions.
+- Express is *not opinionated*, meaning Express does not enforce any "right way" of doing things. You can use any compatible middleware, and you can structure the app as you wish, making it flexible.
+- We can integrate with a [template rendering engine](https://www.digitalocean.com/community/tutorials/nodejs-express-template-engines)(also called as view rendering engine in some articles.) of our choice like Jade, Pug, EJS, etc. A template engine enables you to use static template files and at runtime change the values of variables in those files.
```
Comment (louisefindlay23) [inline articles/express/index.md:40]: Remove full stop after articles

File: articles/express/index.md [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+  });
+```
+
+The code above creates a basic Express application. To run this script, go to your command prompt and enter the command `node app.js` in the project directory. In the console, we can see `Application started and Listening on port 3000` and if we visit `http://localhost:3000/` we can see `HELLO WORLD`.
+
+Let's look at what the code above does.
+
+The first line imports the express module. The second line creates an Express application by calling the top-level `express()` function. Our `app` variable (express application) has methods for handling requests and configuring how the application behaves. We can create multiple apps this way, each with their own requests and responses.
+
+Lets examine the code in section two. `app.get()` is a function, called *route definition*, which tells the express app how to handle an HTTP `GET` request to our server. This function takes 2 main parameters, the first is the route or path which is the relative path from the root of the server; the second is a function that is invoked whenever there is a request to that path. In this case, we are listening for `GET` requests to `/` which is the root of the website.
```
Comment (louisefindlay23) [inline articles/express/index.md:74]: **2** main parameters should be **two**

File: articles/express/index.md [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+
+The second parameter, the callback function, has two arguments `req` and `res`. `req` represents the **request** sent from the browser to the server. `res` represents the **response** that the server sends back.
+
+The code in section three starts a server on the port 3000. You can go to `localhost:3000` to view your response.
+
+### Core parts of Express
+
+#### Middleware
+
+Middleware is a set of functions that sit between a raw request and the final intended route. Middleware functions have access to *all* the HTTP requests coming to the server. Middleware can handle tasks such as logging, sending static files, authorization, session management, etc.
```
Comment (louisefindlay23) [inline articles/express/index.md:84]: Replace comma before session management with and.

File: articles/express/index.md [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+
+The code in section three starts a server on the port 3000. You can go to `localhost:3000` to view your response.
+
+### Core parts of Express
+
+#### Middleware
+
+Middleware is a set of functions that sit between a raw request and the final intended route. Middleware functions have access to *all* the HTTP requests coming to the server. Middleware can handle tasks such as logging, sending static files, authorization, session management, etc.  
+
+In Node.js the request and response objects are passed to one function (request handler) that we write, in Express these objects are passed through a set of functions, called the **middleware stack**. Express will start at the first function in the stack and execute in order down the stack.
```
Comment (louisefindlay23) [inline articles/express/index.md:86]: Add comma after Node.js.

File: articles/express/index.md [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+...
+app.get(/^\/products\/(\d+)$/, function(req, res) {
+    var productId = parseInt(req.params[0], 10);
+    ...
+});
+...
+```
+
+### Template Engines
+Websites are built with HTML, you can dynamically generate HTML pages using Express. Dynamically generated HTML pages are useful when you want to show real time data or change a page's details based on the user. A template engine allows you to use static template files and at runtime replace variables in a template file with actual data. There are different template engines available like [Pug](https://pugjs.org/), Jade, and [EJS](https://ejs.co/). Lets see a basic template using EJS.
```
Comment (louisefindlay23) [inline articles/express/index.md:251]: Lets should have an apostrophe

File: articles/express/index.md [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+
+First let's install it using npm. `npm install ejs` and create a directory called `views` to store your templates and HTML files.
```
Comment (louisefindlay23) [inline articles/express/index.md:253]: Add a qualifier before `npm install ejs` such as "Type" and then add "then" before create to make this paragraph flow better.

File: articles/express/index.md [hunk 18]
```diff
@@ -0,0 +1,318 @@
...
+### Basic Express App
+Install express via npm
+
+```bash
+$ npm install express --save
+```
+
+Let's create a basic Express example app.
+
+```JavaScript
```
Comment (louisefindlay23) [inline articles/express/index.md:52]: This should be javascript or js. Correct other examples.

File: articles/express/index.md [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+
+var app = express();
+...
+app.use(logger_morgan("short")); // logs short notation of requests
+...
+```
+
+Express comes with **`express.static`** middleware bundled with it, it can be used to serve static files instead of the function in the previous section. It provides better security and performance than the function we wrote.
+
+`JavaScript
```
Comment (louisefindlay23) [inline articles/express/index.md:178]: This code block should be surrounded by three backticks. Please correct other examples.

**Review by louisefindlay23 — CHANGES_REQUESTED** [review]
Thanks for making the changes, @rohanreddych. There's some code blocks where you need to fix the markdown formatting which I've pointed out.

File: articles/express/index.md (hunk 19 above)
Comment (rohanreddych) [reply articles/express/index.md:178]: I think it was changed during preliminary review by @hectorkambow . I always type `javascript` but I see `JavaScript` now.

File: articles/express/index.md (hunk 18 above)
Comment (rohanreddych) [reply articles/express/index.md:52]: >I think it was changed during preliminary review by @hectorkambow . I always type javascript but I see JavaScript now.

But I guess `JavaScript` works too.

File: articles/express/index.md (hunk 10 above)
Comment (louisefindlay23) [reply articles/express/index.md]: @rohanreddych, you still need to modify this. 
Please add the word but where I've specified so the sentence makes sense.

Also, you should replace the last comma in the list with and.

**Comment by rohanreddych** [comment]
@louisefindlay23 it says "Your pull request was reviewed 10 hrs ago" but I can still see the yellow dot and it doesn't show anything in the conversation section. Did you request any changes?

**Review by louisefindlay23 — CHANGES_REQUESTED** [review]
@rohanreddych, I replied to a previous comment regarding the first paragraph. See below:

> @rohanreddych, you still need to modify this. 
> Please add the word but where I've specified so the sentence makes sense.
> 
> Also, you should replace the last comma in the list with and.
>

File: articles/express/index.md (hunk 10 above)
Comment (rohanreddych) [reply articles/express/index.md]: >Node.js is fast, portable and written in JavaScript **but** Node.js does not directly support comm ... 
is this correct?

File: articles/express/index.md [hunk 20]
```diff
@@ -0,0 +1,360 @@
...
+Install express via npm
+
+```bash
+$ npm install express --save
+```
+
+Let's create a basic Express example app.
+
+```JavaScript
+//app.js
```
Comment (louisefindlay23) [inline articles/express/index.md:61]: You need to tell readers you need to create the app.js file. Beginners may get stuck here.

File: articles/express/index.md [hunk 21]
```diff
@@ -0,0 +1,360 @@
...
+
+Middleware is a set of functions that sit between a raw request and the final intended route. Middleware functions have access to *all* the HTTP requests coming to the server. Middleware can handle tasks such as logging, sending static files, authorization and session management, etc.  
+
+In Node.js, the request and response objects are passed to one function (request handler) that we write, in Express these objects are passed through a set of functions, called the **middleware stack**. 
+
+Express will start at the first function in the stack and execute in order down the stack.
+
+Every function in the stack takes three arguments `request`, `response` and `next`. `next` is a function, that when called Express executes the next function in the stack. This is a subtle difference between middleware and a route handler which we saw above.
+
+Let's look at a basic static file server to understand middleware. Initialize an npm project and install express. Create a directory named `static` and copy-paste *any* available static files into the folder (text, images, etc.).
```
Comment (louisefindlay23) [inline articles/express/index.md:108]: You've already told readers to install Express earlier in the article. Best practise would be to continue on from the first code snippet.

File: articles/express/index.md [hunk 22]
```diff
@@ -0,0 +1,360 @@
...
+
+We can call the `use()` on the Express app object to add the middleware to the stack.
+
+```JavaScript
+
+var express = require("express");
+var logger_morgan = require("morgan");
+
+var app = express();
+...
```
Comment (louisefindlay23) [inline articles/express/index.md:197]: Both of these three dots should be removed so the code can run

File: articles/express/index.md [hunk 23]
```diff
@@ -0,0 +1,360 @@
...
+app.use(logger_morgan("short")); // logs short notation of requests
+...
+```
+
+Express comes with **`express.static`** middleware bundled with it, it can be used to serve static files instead of the function in the previous section. It provides better security and performance than the function we wrote.
+
+```JavaScript
+app.use(express.static("static") //relative path
```
Comment (louisefindlay23) [inline articles/express/index.md:205]: Add a semicolon after ) for best practise

File: articles/express/index.md [hunk 24]
```diff
@@ -0,0 +1,360 @@
...
+```
+
+Any requested files in the directory "static" are served. `localhost:3000/dummy_file.txt` will show the same result as above. We can call `static()` multiple times. 
+
+If the file is not in the one directory, then the request is passed onto the next `static()` call. When your app gets a request like `/hello.html`, it returns the hello.html file in the public directory.
```
Comment (louisefindlay23) [inline articles/express/index.md:210]: This is a bit confusing. I assume not in the one directory means specified directory? Please reword this to be clearer. 

Also, please explain more about hello.html and the public directory as it relates to the example code as it's also unclear.

File: articles/express/index.md [hunk 25]
```diff
@@ -0,0 +1,360 @@
...
+### Routing
+Express makes request handling easier by mapping requests to different request handlers. A request handler is a function which handles all the requests to a specific path with a specific HTTP method. 
+
+In the basic example above, we saw how to handle a `GET` request. As an application grows in size the routes grow as well as do the request handlers.
+
+Lets see how we can use [Routers](http://expressjs.com/en/4x/api.html#router) to split a large app into smaller, maintainable functions. 
+
+According to the documentation, a Router is "an isolated instance of middleware and routes. Routers can be thought of as “mini” applications only capable of performing middleware and routing".
+
+Routers can be used like middleware functions, they can be "`.use()`d". A simple example:
```
Comment (louisefindlay23) [inline articles/express/index.md:221]: Please rephrase `"`.use()`d"` as it may confuse beginners.

File: articles/express/index.md [hunk 26]
```diff
@@ -0,0 +1,360 @@
...
+
+var app = express();
+
+app.use("/api", apiRouter);
+
+app.listen(3000);
+```
+
+```JavaScript
+//routes/api_router.js
```
Comment (louisefindlay23) [inline articles/express/index.md:237]: Please specify to the reader they have to create the routes folder and api_router.js file because they may gloss over the comments.

File: articles/express/index.md [hunk 27]
```diff
@@ -0,0 +1,360 @@
...
+```JavaScript
+var express = require("express");
+var app = express();
+
+//use a colon to grab a parameter
+
+app.get("/product/:productId", function(req, res){
+    var pid = parseInt(req.params.userid, 10);
+    //res.send   manipulate string to get file with name as productID or something and use a static file server
+    };
```
Comment (louisefindlay23) [inline articles/express/index.md:268]: Linting flags that this line should be });

File: articles/express/index.md [hunk 28]
```diff
@@ -0,0 +1,360 @@
...
+```
+
+**Using Regular Expressions to match routes**
+
+[Regular Expressions (RE)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions) are patterns used to match character combinations in strings. We can use RE to match parameters and define our routes. 
+
+For example, using the example above if we wanted the productId to be only an integer we can try the following:
+
+```JavaScript
+...
```
Comment (louisefindlay23) [inline articles/express/index.md:278]: Again, these dots should be removed

File: articles/express/index.md [hunk 29]
```diff
@@ -0,0 +1,360 @@
...
+app.get(/^\/products\/(\d+)$/, function(req, res) {
+    var productId = parseInt(req.params[0], 10);
+    ...
+});
```
Comment (louisefindlay23) [inline articles/express/index.md:282]: It would be nice if those code snippets produced something the reader could run and see in the browser to see the results.

File: articles/express/index.md [hunk 30]
```diff
@@ -0,0 +1,360 @@
...
+app.set("view engine", "ejs"); //set view engine to ejs
+app.set("views", "views");     //set views directory
+
+app.get("/", function(req, res){   //res.render() renders a view and send the HTML to the client
+        res.render("index", {
+                message: "Hello and Welcome !!!" // this can be any thing you want
+        });
+});
+
+app.listen(3000);
```
Comment (louisefindlay23) [inline articles/express/index.md:308]: This results in `app is not defined`. You need to add `app = express();`

**Review by louisefindlay23 — CHANGES_REQUESTED** [review]
@rohanreddych, I reviewed your code and there's some areas you need to clarify and one of your code snippets is missing a line.

**Comment by rohanreddych** [comment]
@louisefindlay23 any improvements?

File: articles/express/index.md (hunk 29 above)
Comment (louisefindlay23) [reply articles/express/index.md:294]: @rohanreddych, I receive the console message but get this error afterwards: `TypeError: path must be absolute or specify root to res.sendFile`

File: articles/express/index.md [hunk 31]
```diff
@@ -0,0 +1,373 @@
...
+var express = require("express");
+var app = express();
+
+//use a colon to grab a parameter
+
+app.get("/product/:productId", function(req, res){
+    var pid = parseInt(req.params.userid, 10);
+    //res.send   manipulate string to get file with name as productID or something and use a static file server
+    });
+app.listen(3000);
```
Comment (louisefindlay23) [inline articles/express/index.md:275]: Should the user run this code or is it just an example. I can't get an output from it.

**Review by louisefindlay23 — CHANGES_REQUESTED** [review]
@rohanreddych, thanks for making the changes. There's an error produced in one of your code snippets which should be fixed before I can approve your article for final review.

File: articles/express/index.md (hunk 31 above)
Comment (rohanreddych) [reply articles/express/index.md:280]: @louisefindlay23 I didn't write any res.send methods so you cannot see anything.

Comment (rohanreddych) [reply articles/express/index.md:280]: >@rohanreddych, I receive the console message but get this error afterwards: TypeError: path must be absolute or specify root to res.sendFile

this is an example code you cannot use this "as is", it is supposed to give a basic idea

Comment (rohanreddych) [reply articles/express/index.md:280]: I can write the full executable code in a "gist" and share it is ok.

Comment (louisefindlay23) [reply articles/express/index.md:280]: Yes, that would be great @rohanreddych.

**Comment by rohanreddych** [comment]
@louisefindlay23  You can examine the code [here](https://gist.github.com/rohanreddych/137c59f7e57edbfbc839dcd440f0daeb), I gave a link to this in the article.

**Review by louisefindlay23 — APPROVED** [review]
Great work, @rohanreddych. Thanks for providing the extended code. Ready for final review, @hectorkambow.

**Comment by hectorkambow** [comment]
Looking over for Final review now @louisefindlay23  @rohanreddych  
Thank you for the review Louise - 
Rohan - please look over the suggestions and edits you made to better incorporate these types of edits before submitting - to cut down review time and the # of revisions 🚀
